File: src/macos/command.rs
use log::debug;
use super::{EventContext, Key, Flags, MouseButton};

#[derive(Debug)]
pub enum Command {
    Null,
    MouseMoveTo(i32, i32),
    MouseMoveRelative(i32, i32),
    MouseDown(MouseButton),
    MouseUp(MouseButton),
    MouseClick(MouseButton),
    MouseNthClick(MouseButton, u32),
    MouseScrollX(i32),
    MouseScrollY(i32),
    KeyDown(Key),
    KeyUp(Key),
    KeyClick(Key),
    KeyClickFlags(Key, Flags),
}

impl EventContext {
    pub fn evaluate_command(&mut self, command: Command) {
        debug!("{:?}", command);

        use Command::*;

        match command {
            Null => {},
            MouseMoveTo(x, y) => self.mouse_move_to(x, y),
            MouseMoveRelative(x, y) => self.mouse_move_relative(x, y),
            MouseDown(button) => self.mouse_down(button),
            MouseUp(button) => self.mouse_up(button),
            MouseClick(button) => self.mouse_click(button),
            MouseNthClick(button, count) => self.mouse_nth_click(button, count),
            MouseScrollX(length) => self.mouse_scroll_x(length),
            MouseScrollY(length) => self.mouse_scroll_y(length),
            KeyDown(key) => self.key_down(key),
            KeyUp(key) => self.key_up(key),
            KeyClick(key) => self.key_click(key),
            KeyClickFlags(key, flags) => self.key_click_flags(key, flags),
        }
    }
}

File: Remote/Remote/Constants/Command.swift
//
//  Command.swift
//  Remote
//
//  Created by Indiana Kernick on 9/2/21.
//  Copyright © 2021 Indiana Kernick. All rights reserved.
//

import Foundation

fileprivate func setInt(_ buf: inout Data, index: Int, value: Int16) {
    buf[index] = UInt8((value >> 8) & 0xFF)
    buf[index + 1] = UInt8(value & 0xFF)
}

enum Command {
    static func keyDown(_ key: Key, repeat: Bool = false) -> Data {
        Data([CommandCode.keyDown.rawValue, key.rawValue, `repeat` ? 1 : 0])
    }
    
    static func keyUp(_ key: Key) -> Data {
        Data([CommandCode.keyUp.rawValue, key.rawValue])
    }
    
    static func keyClick(_ key: Key) -> Data {
        Data([CommandCode.keyClick.rawValue, key.rawValue])
    }
    
    static func keyClick(_ key: Key, with modifier: Key) -> Data {
        Data([
            CommandCode.keyDown.rawValue, modifier.rawValue, 0,
            CommandCode.keyClick.rawValue, key.rawValue,
            CommandCode.keyUp.rawValue, modifier.rawValue
        ])
    }
    
    static func mouseMoveRel() -> Data {
        Data([CommandCode.mouseMoveRel.rawValue, 0, 0, 0, 0])
    }
    
    static func mouseMoveAbs() -> Data {
        Data([CommandCode.mouseMoveAbs.rawValue, 0, 0, 0, 0])
    }
    
    static func mouseWarp() -> Data {
        Data([CommandCode.mouseWarp.rawValue, 0, 0, 0, 0])
    }
    
    static func mouseScroll() -> Data {
        Data([CommandCode.mouseScroll.rawValue, 0, 0, 0, 0])
    }
    
    static func setMouseParams(_ buf: inout Data, x: Int16, y: Int16) {
        setInt(&buf, index: 1, value: x);
        setInt(&buf, index: 3, value: y);
    }
    
    static func mouseDown(_ button: MouseButton, count: UInt8 = 1) -> Data {
        Data([CommandCode.mouseDown.rawValue, button.rawValue, count])
    }
    
    static func mouseUp(_ button: MouseButton, count: UInt8 = 1) -> Data {
        Data([CommandCode.mouseUp.rawValue, button.rawValue, count])
    }
    
    static func mouseClick(_ button: MouseButton, count: UInt8 = 1) -> Data {
        Data([CommandCode.mouseClick.rawValue, button.rawValue, count])
    }
}
